disciple-tools-data-reporting-provider-azure.php

```diff
@@ -237,12 +256,225 @@ public function data_reporting_export( $columns, $rows, $type, $config ) {
+        if ( empty( $storage_account_key ) ) {
+            echo "<p>A Storage Account Key has not been set. Please update in <a href='$settings_link'>Settings</a></p>";
+        } else if ( empty( $storage_account ) ) {
+            echo "<p>A Storage Account has not been set. Please update in <a href='$settings_link'>Settings</a></p>";
+        } else if ( empty( $storage_account_container ) ) {
+            echo "<p>A Storage Account Container has not been set. Please update in <a href='$settings_link'>Settings</a></p>";
+        } else {
+            $columns = array_map(function ( $column ) { return $column['name']; }, $columns);
+            // TODO: do not hardcode maxmemory value
+            $csv = fopen('php://temp/maxmemory:'. (100*1024*1024), 'r+');
+            fputcsv($csv, $columns);
+            // loop over the rows, outputting them
+            foreach ($rows as $row ) {
+                fputcsv( $csv, $row );
+            }
+            rewind($csv);
+            $content = stream_get_contents($csv);
+            // Azure specifics
+            $blob_name = "contacts_".strval(gmdate('Ymdhi', time())).".csv";
+            $connectionString = "DefaultEndpointsProtocol=https;AccountName=".$storage_account.";AccountKey=".$storage_account_key.";EndpointSuffix=core.windows.net";
+            $blobClient = MicrosoftAzure\Storage\Blob\BlobRestProxy::createBlobService($connectionString);
+            // TODO: RBAC Support
+            //$aadtoken = "";
+            //$blobClient = MicrosoftAzure\Storage\Blob\BlobRestProxy::createBlobServiceWithTokenCredential($aadtoken, $connectionString);
+            try {
+              //Upload blob
+              $blobClient->createBlockBlob($storage_account_container, $blob_name, $content);
+              echo '<div class="notice notice-success notice-dt-data-reporting is-dismissible" data-notice="dt-data-reporting"><p>'.$blob_name.' successfully uploaded to Azure Blob</p></div>';
+            } catch(MicrosoftAzure\Storage\Common\Exceptions\ServiceException $e){
+              $code = $e->getCode();
+              $error_message = $e->getMessage();
+              echo $code.": ".$error_message."<br />";
+            }
+        }
     }
 
     public function data_reporting_tab( ) {
       ?>
-      <h2>My Azure Provider</h2>
-      <p>Add here any getting started or how-to information that is needed for your provider</p>
+        <script>
+          // see: https://www.30secondsofcode.org/blog/s/copy-text-to-clipboard-with-javascript
+          function copyText(elementId) {
+            const el = document.createElement('textarea');
+            el.value = document.getElementById(elementId).value;
+            el.setAttribute('readonly', '');
+            el.style.position = 'absolute';
+            el.style.left = '-9999px';
+            document.body.appendChild(el);
+            el.select();
+            document.execCommand('copy');
+            document.body.removeChild(el);
+          }
```

Comment:
@zdmc23 I'm guessing this will be useful for other providers that are created too, so if you want to submit it to the main plugin so that it's available in the future, feel free to do so. Same thing for the `.pre-div` style above.